### PR TITLE
Isotonic Regression

### PR DESCRIPTION
**Pull request by wildart**
An isotonic regression  based the pool adjacent violators algorithm from *Best, M.J., Chakravarti, N. Active set algorithms for isotonic regression; A unifying framework. Mathematical Programming 47, 425–439 (1990)*.

### CODE REVIEW

File: algorithms/linfa-linear/src/isotonic.rs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+pub struct FittedIsotonicRegression<F> {
+    regressor: Array1<F>,
+    response: Array1<F>,
+}
+
+impl Default for IsotonicRegression {
+    fn default() -> Self {
+        IsotonicRegression::new()
+    }
+}
```
Comment (YuhanLiin) [inline algorithms/linfa-linear/src/isotonic.rs:50]: This can just be derived

File: algorithms/linfa-linear/src/isotonic.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        let y = dataset.as_single_targets();
+
+        // Check the input dimension
+        assert_eq!(dim, 1, "The input dimension must be 1.");
+
+        // Check that our inputs have compatible shapes
+        assert_eq!(y.dim(), n_samples);
+
+        // use correlation for determining relationship between x & y
+        let x = X.slice(s![.., 0]);
```
Comment (YuhanLiin) [inline algorithms/linfa-linear/src/isotonic.rs:87]: Do `X.col(0)`

File: algorithms/linfa-linear/src/isotonic.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        let rho = DatasetBase::from(stack![Axis(1), x, y]).pearson_correlation();
+        let incresing = rho.get_coeffs()[0] >= F::zero();
```
Comment (YuhanLiin) [inline algorithms/linfa-linear/src/isotonic.rs:89]: `increasing`

File: algorithms/linfa-linear/src/isotonic.rs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+            let B_plus = &J[i + 1];
+            if B_zero.0 <= B_plus.0 {
+                i += 1;
+            } else {
+                let w0 = B_zero.1 + B_plus.1;
+                let v0 = (B_zero.0 * B_zero.1 + B_plus.0 * B_plus.1) / w0;
+                let mut i0 = B_zero.2.clone();
+                i0.extend(&(B_plus.2));
+                J[i] = (v0, w0, i0);
+                J.remove(i + 1);
```
Comment (YuhanLiin) [inline algorithms/linfa-linear/src/isotonic.rs:115]: Vector removal is O(n), which is inefficient inside loops. Can you change `J` to a different data structure so removals are less expensive? `Hashmap` or `Vec<Option<...>>` are good options, but you'll need to change the way you index into `J`.

**Comment by wildart** [comment]
I rewrote algorithm, and added an another implementation.

**Comment by YuhanLiin** [comment]
What's the difference and tradeoffs between PVA and AlgorithmA? Is there a reason why we want to include both instead of just including one?

**Comment by wildart** [comment]
I removed AlgorithmA code, and left PVA code only. Both algorithms are of *O(n)* complexity, it's just AlgorithmA is dual to PVA, so the solutions they provide are almost similar.

After all, I do not think we need two algorithms implemented.

**Comment by YuhanLiin** [comment]
Mention the paper/textbook the algorithm is from in the public docs and you're good to go

**Review by YuhanLiin — APPROVED** [review]

**Pull request merged**